### PR TITLE
Outputs the line number from where the logging request was issued

### PR DESCRIPTION
**Pull request by chaosky**
Motivation:

Outputs the line number from where the logging request was issued when debugging netty

Modification:

Add LocationAwareSlf4JLogger can output the correct line number

Result:

Users of this library can correctly output the line number of the log when using slf4j

### CODE REVIEW

File: common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java [hunk 1]
```diff
@@ -34,6 +36,8 @@
 public abstract class InternalLoggerFactory {
 
     private static volatile InternalLoggerFactory defaultFactory;
+    static final String LINE_NUMBER = "io.netty.loggerLineNumber";
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java]: @chaosky also should this be something that has `slf4j` in the name as its specific to slf4j ?

File: common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java [hunk 2]
```diff
@@ -53,6 +57,13 @@ private static InternalLoggerFactory newDefaultFactory(String name) {
         return f;
     }
 
+
+    static {
+        boolean lineNumber = SystemPropertyUtil.getBoolean(LINE_NUMBER, false);
+        LOGGER_LINE_NUMBER = lineNumber;
+        getDefaultFactory().newInstance(InternalLoggerFactory.class.getName()).debug("-D{}: {}", LINE_NUMBER, lineNumber);
```
[the remark this answers is not in the document]
Comment (chaosky) [reply common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java]: I think in most cases, the need to output the line number occurs when learning netty or when debugging netty, so a system property switch is provided. In other cases, the basic Logger is enough.

My ability only implements slf4j's functionality, so a LINE_NUMBER is provided for other implementations (such as Log4JLogger, Log4J2Logger output line numbers) to use

File: common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java (hunk 1 above)
Comment (chaosky) [reply common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java]: Sorry, I have signed CLA.

File: common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chaosky) [reply common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java]: I have no idea how to test it(line number of log), so this feature is off by default

**Comment by normanmaurer** [comment]
@chaosky sorry but can you explain me in detail why this PR is needed ?

**Comment by chaosky** [comment]
@normanmaurer  The PR same as [7185](https://github.com/netty/netty/pull/7185) , but it is adapted slf4j.

**Comment by normanmaurer** [comment]
@chaosky ah got it!

File: common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java (hunk 1 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java]: @chaosky did you see the comment ?

File: common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+package io.netty.util.internal.logging;
+
+import org.slf4j.Marker;
+import org.slf4j.spi.LocationAwareLogger;
+
+import static org.slf4j.spi.LocationAwareLogger.*;
+
+/**
+ * <a href="http://www.slf4j.org/">SLF4J</a> logger.
+ * similar to jcl-over-slf4j/SLF4JLocationAwareLog.java
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java]: nit: use uppercase letter to start sentence and also use `{@link }` to link to other classes.

File: common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java [hunk 4]
```diff
@@ -53,6 +57,13 @@ private static InternalLoggerFactory newDefaultFactory(String name) {
         return f;
     }
 
+    static {
+        boolean lineNumber = SystemPropertyUtil.getBoolean(LINE_NUMBER, false);
+        LOGGER_LINE_NUMBER = lineNumber;
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java]: just assign directly ?

**Comment by normanmaurer** [comment]
@chaosky ping

**Comment by normanmaurer** [comment]
closing in favour of https://github.com/netty/netty/pull/8258. Thanks!

**Pull request closed without merging**